src/miniscript/analyzable.rs

```diff
@@ -0,0 +1,151 @@
+                f.write_str("Atleast one spend path exceeds the resource limits(stack depth/satisfaction size..)")
+            }
+            AnalysisError::HeightTimeLockCombination => {
+                f.write_str("Contains a combination of heightlock and timelock")
+            }
+            AnalysisError::Malleable => f.write_str("Miniscript is malleable")
+        }
+    }
+}
+
+impl error::Error for AnalysisError {
+    fn description(&self) -> &str {
+        ""
+    }
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> Miniscript<Pk, Ctx> {
+    /// Whether all spend paths of miniscript require a signature
+    pub fn requires_sig(&self) -> bool {
+        self.ty.mall.safe
+    }
+
+    /// Whether the miniscript is malleable
+    pub fn is_non_malleable(&self) -> bool {
+        self.ty.mall.non_malleable
+    }
+
+    /// Whether the miniscript can exceed the resource limits(Opcodes, Stack limit etc)
+    // It maybe possible to return a detail error type containing why the miniscript
+    // failed. But doing so may require returning a collection of errors
+    pub fn within_resource_limits(&self) -> bool {
+        match Ctx::check_satisfaction_rules(&self) {
+            Ok(_) => true,
+            Err(_) => false,
+        }
+    }
+
+    /// Whether the miniscript contains a combination of timelocks
+    pub fn has_mixed_timelocks(&self) -> bool {
+        self.ext.timelock_info.contains_unspendable_path()
+    }
+
+    /// Whether the miniscript has repeated Pk or Pkh
+    pub fn has_repeated_keys(&self) -> bool {
+        // Simple way to check whether all of these are correct is
+        // to have an iterator
+        let all_pkhs_len = self
+            .iter_pk_pkh()
+            .map(|pk_pkh| match pk_pkh {
+                PkPkh::PlainPubkey(pk) => pk.to_pubkeyhash(),
+                PkPkh::HashedPubkey(h) => h,
+            })
+            .count();
+
+        let unique_pkhs_len = self
+            .iter_pk_pkh()
+            .map(|pk_pkh| match pk_pkh {
+                PkPkh::PlainPubkey(pk) => pk.to_pubkeyhash(),
+                PkPkh::HashedPubkey(h) => h,
+            })
+            .collect::<HashSet<_>>()
+            .len();
+
+        unique_pkhs_len != all_pkhs_len
+    }
+
+    /// Check whether the underlying Miniscript is safe under the current context
+    /// Lifting these polices would create a semantic representation that does
+    /// not represent the underlying semantics when miniscript is spent.
+    /// Signing logic may not find satisfaction even if one exists.  
+    ///
+    /// For most cases, users should be dealing with safe scripts.
+    /// Use this function to check whether the guarantees of library hold.
+    /// Most functions of the library like would still
+    /// work, but results cannot be relied upon
+    pub fn check_safety(&self) -> Result<(), AnalysisError> {
```

Comment:
should rename

Reply:
Maybe `sanity_check`

And then we could have `parse_insane` and `from_str_insane` :P